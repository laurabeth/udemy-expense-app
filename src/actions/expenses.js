import uuid from "uuid";

export const addExpense = ({
  amount = 0,
  createdOn = 0,
  description = "",
  note = "",
} = {}) => {
  const id = uuid();

  return {
    expense: {
      amount: amount,
      createdOn: createdOn,
      description: description,
      id,
      note: note,
    },
    type: "ADD_EXPENSE",
  };
};

export const removeExpense = ({ id } = {}) => {
  return {
    id,
    type: "REMOVE_EXPENSE",
  };
};

export const editExpense = (id, updates) => {
  return {
    id,
    type: "EDIT_EXPENSE",
    updates,
  };
};
