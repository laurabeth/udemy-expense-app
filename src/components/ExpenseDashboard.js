import React from "react";

const ExpenseDashboard = () => <div> This is the dashboard component </div>;

export default ExpenseDashboard;
