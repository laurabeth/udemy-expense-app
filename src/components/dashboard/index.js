export { default as ExpenseDashboard } from "./ExpenseDashboard";
export { default as ExpenseList } from "./ExpenseList";
export { default as ExpenseListFilters } from "./ExpenseListFilters";
export { default as ExpenseListItem } from "./ExpenseListItem";
