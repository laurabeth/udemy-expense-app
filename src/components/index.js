export * from "./dashboard";
export { default as AddExpense } from "./AddExpense";
export { default as EditExpense } from "./EditExpense";
export { default as ExpenseForm } from "./ExpenseForm";
export { default as Header } from "./Header";
export { default as Help } from "./Help";
export { default as NotFound } from "./NotFound";
