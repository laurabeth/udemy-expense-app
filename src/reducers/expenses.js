const defaultExpenses = [];

const expensesReducer = (state = defaultExpenses, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        return expense.id === action.id ? { ...expense, ...action.updates } : expense;
      });
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "SET_EXPENSES":
      return action.expenses;
    default:
      return state;
  }
};

export default expensesReducer;
