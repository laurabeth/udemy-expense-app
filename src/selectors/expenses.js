const getVisibleExpenses = (expenses, filters) => {
  const { text, sortBy, startDate, endDate } = filters;

  return expenses
    .filter((expense) => {
      const startDateMatch = typeof startDate !== "number" || expense.createdOn >= startDate;
      const endDateMatch = typeof endDate !== "number" || expense.createdOn <= endDate;
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdOn < b.createdOn ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
