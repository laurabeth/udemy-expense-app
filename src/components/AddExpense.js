import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

export const AddExpense = ({ history, addExpense }) => {
  const submitExpense = (expense) => {
    addExpense(expense);
    history.push("/");
  };

  return (
    <>
      <h1>AddExpense</h1>
      <ExpenseForm submitExpense={submitExpense} />
    </>
  );
};

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpense);
