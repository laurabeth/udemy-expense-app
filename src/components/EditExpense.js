import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions";

export const EditExpense = ({ editExpense, expense, history, removeExpense }) => {
  const submitEdit = (edits) => {
    editExpense(expense.id, edits);
    history.push("/");
  };
  const handleRemoveExpense = () => {
    removeExpense(expense.id);
    history.push("/");
  };
  return (
    <div>
      <ExpenseForm expense={expense} submitExpense={submitEdit} />
      <button onClick={handleRemoveExpense}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
  };
};

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, edits) => dispatch(editExpense(id, edits)),
  removeExpense: (id) => dispatch(removeExpense({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
