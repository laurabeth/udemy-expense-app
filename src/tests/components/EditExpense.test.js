/* eslint-disable no-undef */
import React from "react";
import { shallow } from "enzyme";
import { EditExpense } from "../../components/EditExpense";
import expenses from "../fixtures/expenses";

let editExpense, history, removeExpense, wrapper;
beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpense
      editExpense={editExpense}
      expense={expenses[2]}
      history={history}
      removeExpense={removeExpense}
    />,
  );
});

describe("edit expense", () => {
  it("renders", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("handles submit edit", () => {
    const edits = { note: "test note" };
    wrapper.find("ExpenseForm").prop("submitExpense")(edits);
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, edits);
  });

  it("handles remove expense", () => {
    wrapper.find("button").prop("onClick")();
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id);
  });
});
