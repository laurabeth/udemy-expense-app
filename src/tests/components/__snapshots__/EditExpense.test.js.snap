// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit expense renders 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 800,
        "createdOn": "1970-01-02T00:00:00.000Z",
        "description": "chocolate",
        "id": "3",
        "note": "",
      }
    }
    submitExpense={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
